crates/red_knot_python_semantic/src/types/type_ordering.rs

```diff
@@ -11,16 +11,22 @@ use super::{
 /// in an [`crate::types::IntersectionType`] or a [`crate::types::UnionType`] in order for them
 /// to be compared for equivalence.
 ///
-/// Two unions with equal sets of elements will only compare equal if they have their element sets
-/// ordered the same way.
+/// Two intersections are compared lexicographically.
+/// Two unions are never compared in this function because DNF does not permit nested unions.
+///
+/// Element types must already be sorted.
 ///
 /// ## Why not just implement [`Ord`] on [`Type`]?
 ///
 /// It would be fairly easy to slap `#[derive(PartialOrd, Ord)]` on [`Type`], and the ordering we
 /// create here is not user-facing. However, it doesn't really "make sense" for `Type` to implement
 /// [`Ord`] in terms of the semantics. There are many different ways in which you could plausibly
 /// sort a list of types; this is only one (somewhat arbitrary, at times) possible ordering.
-pub(super) fn union_elements_ordering<'db>(left: &Type<'db>, right: &Type<'db>) -> Ordering {
+pub(super) fn union_or_intersection_elements_ordering<'db>(
+    left: &Type<'db>,
+    right: &Type<'db>,
+    db: &'db dyn crate::Db,
```

Comment:
nit: for all other functions and methods in red-knot, `db` is always the first argument. We should do that here too, for consistency:

```suggestion
    db: &'db dyn crate::Db,
    left: &Type<'db>,
    right: &Type<'db>,
```